Sources/NIOCore/ByteBuffer-core.swift

```diff
@@ -1119,3 +1117,100 @@ extension ByteBuffer {
+    /// in a format that's compatible with `xxd -r -p`.
+    ///
+    /// - parameters:
+    ///     - offset:  The offset from the beginning of the buffer, in bytes. Defaults to 0.
+    ///     - readableOnly: Whether to only dump the `readableBytes` part (before the `writerIndex`), or to dump the full buffer, including empty bytes in the end of the allocated space. Defaults to true.
+    func hexDumpShort(offset: Int = 0, readableOnly: Bool = true) -> String {
+        let length = (readableOnly ? self.writerIndex : self.capacity) - offset
+        return self._storage.dumpBytes(slice: self._storage.fullSlice,
+                                       offset: offset,
+                                       length: length)
+    }
+
+    /// Return a `String` of hexadecimal digits of bytes in the Buffer.,
+    /// in a format that's compatible with `xxd -r -p`, but clips the output to the max length of `limit` bytes.
+    /// If the dump contains more than the `limit` bytes, this function will return the first `limit/2`
+    /// and the last `limit/2` of that, replacing the rest with `...`, i.e. `01 02 03 ... 09 11 12`.
+    ///
+    /// - parameters:
+    ///     - offset:  The offset from the beginning of the buffer, in bytes. Defaults to 0.
+    ///     - readableOnly: Whether to only dump the `readableBytes` part (before the `writerIndex`), or to dump the full buffer,
+    ///         including empty bytes in the end of the allocated space. Defaults to true.
+    ///     - limit: The maximum amount of bytes presented in the dump.
+    func hexDumpShort(offset: Int = 0, readableOnly: Bool = true, limit: Int) -> String {
+        let lastIndex = readableOnly ? self.writerIndex : self.capacity
+        let length = lastIndex - offset
+
+        // If the length of the resulting hex dump is smaller than the maximum allowed dump length,
+        // return the full dump. Otherwise, dump the first and last pieces of the buffer, then concatenate them.
+        if length <= limit {
+            return self.hexDumpShort(offset: offset, readableOnly: readableOnly)
+        } else {
+            let clipLength = limit / 2
+            let startHex = self._storage.dumpBytes(slice: self._storage.fullSlice,
+                                                   offset: offset,
+                                                   length: offset + clipLength)
+
+            let endHex = self._storage.dumpBytes(slice: self._storage.fullSlice,
+                                                 offset: lastIndex - clipLength,
+                                                 length: clipLength)
+
+            return startHex + " ... " + endHex
+        }
+    }
+
+    /// Returns a `String` of hexadecimal digits of bytes in the Buffer,
+    /// with formatting compatible with output of `hexdump -C`.
+    func hexDumpLong(offset: Int = 0, readableOnly: Bool = true) -> String {
+        var result = ""
```

Comment:
While this is a debugging tool it'd be nice if we could reduce allocations where possible. As we know how many bytes are in the buffer and what the output looks like we can estimate how long the result will be and use `result.reserveCapacity(...)` to allocate that space ahead of time (otherwise we end up repeatedly doubling the size of the resulting string as we append to it which can get quite expensive).